File: test/xsd/interations/case2/case2.go
package schema

import (
	"errors"
	"fmt"
)

// "AttributeGroup"
// AgDesc ...
type AgDesc struct {
	DescAttr *string `xml:"desc,attr,omitempty" json:"desc,omitempty"`
}

func (t *AgDesc) Validate() error {

	return nil
}

// "ComplexType"
// TBaseElement ...
type TBaseElement struct {
	NodeID
	AdditionalFields AdditionalFieldsType `xml:",any,attr"`
	CustomElements   []AnyHolder          `xml:",any"`
}

func (t *TBaseElement) Validate() error {
	// AttributeGroup

	// Attributes
	// Elements

	//todo: check "minOccurs" and "maxOccurs"
	return nil
}

// "ComplexType"
// TUnNaming ...
type TUnNaming struct {
	TBaseElement
	AgDesc // AttributeGroup
}

func (t *TUnNaming) Validate() error {
	if err := t.TBaseElement.Validate(); err != nil {
		return err
	}
	// AttributeGroup
	if err := t.AgDesc.Validate(); err != nil {
		return err
	}

	// Attributes
	// Elements

	//todo: check "minOccurs" and "maxOccurs"
	return nil
}

// "ComplexType"
// TNaming ...
type TNaming struct {
	TBaseElement
	AgDesc         // AttributeGroup
	NameAttr TName `xml:"name,attr" json:"name"`
}

func (t *TNaming) Validate() error {
	if err := t.TBaseElement.Validate(); err != nil {
		return err
	}
	// AttributeGroup
	if err := t.AgDesc.Validate(); err != nil {
		return err
	}

	// Attributes
	if err := t.NameAttr.Validate(); err != nil {
		return err
	}
	// check use='required'  type="tName"
	if t.NameAttr == "" {
		return fmt.Errorf("requred attribute %q in %s", "name", "TNaming")
	}
	// Elements

	//todo: check "minOccurs" and "maxOccurs"
	return nil
}

// "SimpleType"
// base:"tAnyName"
// TName ...
type TName string

func (t *TName) Validate() error {
	if len(*t) < 1 {
		return errors.New("меньше минимального 1 значение в TName")
	}

	return nil
}

// "SimpleType"
// base:"xs:normalizedString"
// TAnyName ...
type TAnyName string

func (t *TAnyName) Validate() error {

	return nil
}

File: test/xsd/interations/case2/case2.ts
import { Constructor,NodeID } from "./BaseIndex"


// AgDesc ...
export function AgDesc<TBase extends Constructor>(Base: TBase) {
  return class Scaling extends Base {
    desc?: string;

  constructor(...args: any[]) {
    super(...args);
    if (args[0]?.hasOwnProperty("desc")) this.desc = args[0]['desc'];

  }
  };
}
// Simple type
// TAcsiName ...
export type TAcsiName = string;

// Simple type
// TIEDName ...
export type TIEDName = string;

// AgLDRef ...
export function AgLDRef<TBase extends Constructor>(Base: TBase) {
  return class Scaling extends AgDesc(Base) {
    iedName: string;
  ldInst: boolean;

  constructor(...args: any[]) {
    super(...args);
    if (args[0]?.hasOwnProperty("iedName")) this.iedName = args[0]['iedName'];
    if (args[0]?.hasOwnProperty("ldInst")) this.ldInst = args[0]['ldInst'];

  }
  };
}
// AgLNRef ...
export function AgLNRef<TBase extends Constructor>(Base: TBase) {
  return class Scaling extends AgLDRef(Base) {
    prefix?: string;
  lnClass: string;
  lnInst: string;

  constructor(...args: any[]) {
    super(...args);
    if (args[0]?.hasOwnProperty("prefix")) this.prefix = args[0]['prefix'];
    if (args[0]?.hasOwnProperty("lnClass")) this.lnClass = args[0]['lnClass'];
    if (args[0]?.hasOwnProperty("lnInst")) this.lnInst = args[0]['lnInst'];

  }
  };
}
// Simple type
// TName ...
export type TName = string;

// Simple type
// TAnyName ...
export type TAnyName = string;

// ComplexType 
//  TBaseElement ...
export class TBaseElement extends NodeID {
  constructor(...args: any[]) {
    super(...args);
	const opts = args[0];
    // Attributes
	
    // Elements
    
  }}

// ComplexType 
//  TNaming ...
export class TNaming extends AgDesc(TBaseElement) {
	name: TAnyName;
  constructor(...args: any[]) {
    super(...args);
	const opts = args[0];
    // Attributes
	    this.name = opts['name'];

    // Elements
    
  }}

// ComplexType 
//  TUnNaming ...
export class TUnNaming extends AgDesc(TBaseElement) {
  constructor(...args: any[]) {
    super(...args);
	const opts = args[0];
    // Attributes
	
    // Elements
    
  }}
